[1기-C] 김지훈- JPA 게시판 만들기

DevRunner21 opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
- Spring Data JPA를 활용하여 게시판 구현

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->

### 엔티티를 구성한다
- [x] 엔티티를 구성한다
  - 회원(User)
  - 게시글(Post)

- [x] 회원과 게시글에 대한 연관관계를 설정한다.
  - 회원과 게시글은 1:N 관계이다.
  
- [x] 게시글 Repository를 구현한다. (PostRepository)


### API를 구현한다.
- [x] 게시글 조회
    - 페이징 조회 (GET "/posts")
    - 단건 조회 (GET "/posts/{id}")
- [x] 게시글 작성 (POST "/posts")
- [x] 게시글 수정 (PATCH "/posts/{id}")
- [x] 게시글 삭제 (DELETE "/posts/{id}")

### REST-DOCS를 이용해서 문서화한다.
- [x] REST-DOCS를 이용해서 문서화

## ✅ 피드백 반영사항  <!-- 지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
- 예외처리 방법이 적절하게 되었는지.. 더 효율적인 방법은 없는지..

DevRunner21 committed:
[init] 프로젝트를 생성하라.

DevRunner21 committed:
[init] DB 세팅을 진행하라.

- 개발 : MySql
- 테스트 : H2

DevRunner21 committed:
[style]  설정파일을 수정하라.

DevRunner21 committed:
[FEAT] BaseEntity를 생성하고 Auditing 설정을 진행하라.

DevRunner21 committed:
[feat] Post, User Entity를 생성하고 연관관계를 매핑하라.

DevRunner21 committed:
[feat] PostRepository를 생성하라.

DevRunner21 committed:
[test] PostRepository의 게시글 페이징 조회 기능을 테스트 하라.

DevRunner21 committed:
[feat] PostService 구현에 필요한 예외를 정의하라.

- PostNotFoundException 정의
- UserNotFoundException 정의
- ErrorInfo Enum 정의

DevRunner21 committed:
[feat] Spring Validation 라이브러리를 추가하라.

DevRunner21 committed:
[feat] PostService에 필요한 DTO를 정의하라.

DevRunner21 committed:
[feat] UserRepository를 생성하고 "LoginId로 조회" 메서드를 추가하라.

DevRunner21 committed:
[feat] PostService를 수현하라.

- 페이징 리스트 조회
- 단건조회
- 게시글 등록
- 게시글 수정
- 게시글 삭제

DevRunner21 committed:
[test] PostService의 테스트 코드를 작성하라.

DevRunner21 committed:
[refact] UserNotFoundException, PostNotFoundException을 BusinessException으로 추상화하라.

DevRunner21 committed:
[feat] API 공통 Response 객체를 정의하라.

DevRunner21 committed:
[fix] PostForm 직렬화를 위해 생성자를 추가하라.

DevRunner21 committed:
[fix] 잘못 커밋된 소스를 수정하라.

DevRunner21 committed:
[feat] PostController를 구현하라.

- 페이징 리스트 조회
- 단건조회
- 게시글 생성
- 게시글 수정
- 게시글 삭제

DevRunner21 committed:
[feat] ExceptionHandler를 정의하라.

DevRunner21 committed:
[test] PostController의 테스트 코드를 작성하라.

Young-BLUE commented on devrun/src/main/java/kdt/prgrms/devrun/post/service/PostService.java:
```diff
@@ -0,0 +1,23 @@
+    Long updatePost(Long postId ,PostForm postForm);
+
+    void deletePostById(Long postId);
+
```
PostService를 하나의 class에서 작성하지 않고 interface와 SimplePostService라는 구현 class로 나누신 이유가 궁금합니다!

Young-BLUE commented on devrun/src/main/resources/application.yml:
```diff
@@ -0,0 +1,16 @@
+      hibernate:
+        format_sql: true
+logging.level:
+  org.hibernate.SQL: debug
```
SQl logging을 이용할 수 있군요! 덕분에 좋은 정보 알아갑니다!

Young-BLUE commented on devrun/src/test/java/kdt/prgrms/devrun/post/controller/PostApiControllerTest.java:
```diff
@@ -0,0 +1,125 @@
+            .contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk())
+            .andDo(print());
+
```
api/v1/... 으로 구성하신 이유가 있나요?

DevRunner21 replied:
이 부분은 현호님과 나눈 이야기가 있어서 공유합니다! ㅎㅎ
https://github.com/prgrms-be-devcourse/SpringBoot-Board/pull/5#discussion_r725798207

Young-BLUE replied:
이미 나누신 이야기가 있군요! 버전 관리를 위해 명시해주는 방법이 있었네요 감사합니다 ㅎㅎ

Young-BLUE commented on devrun/src/main/java/kdt/prgrms/devrun/post/dto/SimplePostDto.java:
```diff
@@ -0,0 +1,24 @@
+import javax.validation.constraints.NotBlank;
+import java.time.LocalDateTime;
+
+@Data
```
@Getter와 @Setter 어노테이션만 따로 쓰지 않고 @Data 어노테이션을 쓰면 어떤 장점이 있나요?

DevRunner21 replied:
"@Getter / @Setter, @ToString, @EqualsAndHashCode, @RequiredArgsConstructor " 
위 어노테이션들을 한번에 묶어놓은 것이 "@Data" 어노테이션입니다.

장점으로는 여러줄로 쓸 것을 한번에 처리할 수 있다는 장점이 있죠.

하지만 여러 상황에서 Setter를 제한해야 할 경우도 있고 toString을 재정의해야할 경우가 많기 때문에
보통은 "@Data"를 잘 쓰지는 않지만...

제가 사용한 부분은 따로 뭔가를 재정의하거나 Setter를 제한할 필요가 없는 DTO라서 그냥 사용했습니다.

Young-BLUE replied:
지난번 과제에 저도 @Data를 썼었는데 그당시 코드리뷰 때 Getter/Setter만 필요하다면 다른 어노테이션을 포함할 필요가 있을까? 여서 변경했었는데 이 경우에는 제한할 필요가 없다면 그냥 써도 되겠군요. 그때는 잘 모르고 썼었는데 덕분에 알고 쓸 수 있겠네요 감사합니다 ㅎㅎ

Young-BLUE commented on devrun/src/main/java/kdt/prgrms/devrun/domain/User.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Id
+    @GeneratedValue
+    @Column(name = "user_id")
```
여기에서는 따로 @GeneratedValue 에 strategy가 없는데 저같은 경우에는 SEQUENCE로 지정했습니다. 따로 지정하지 않을 경우에는 어떤 내용이 적용이 되는건가요?

DevRunner21 replied:
아 이 부분 제가 답변하는걸 놓쳤었네요. ㅜㅜ
Auto가 default로 설정되어있기 때문에 그냥 따로 지정하지 않고 사용했었습니다!

Young-BLUE commented on devrun/src/main/java/kdt/prgrms/devrun/domain/User.java:
```diff
@@ -0,0 +1,44 @@
+    private String email;
+
+    private String hobby;
+
```
상당히 많은 유저의 정보들을 구현하셨네요! 저는 요구되는 내용만 구성했었는데 추가로 이런 정보들을 담을 수 있겠군요. 지훈님의 꼼꼼함 배워가겠습니다!

Young-BLUE commented on devrun/src/main/java/kdt/prgrms/devrun/domain/BaseTimeEntity.java:
```diff
@@ -0,0 +1,25 @@
+
+    @LastModifiedDate
+    private LocalDateTime updatedAt;
+
```
BaseEntity로만 구성 할 때와 BaseTimeEntity로 구성 하는 것에 어떠한 차이가 있나요?!
@LastModifiedDate를 이용하면 따로 작성하지 않아도 해당 시간으로 적용이 되는 방식인가요? 처음 보는데 정말 어노테이션에는 무궁무진한 기능들이 있네요!

DevRunner21 replied:
(1) BaseEntity로만 구성 할 때와 BaseTimeEntity로 구성 하는 것에 어떠한 차이가 있나요?!

BaseEntity에는 작성자, 수정자 라는 필드를 가지고 있고
BaseTimeEntity는 작성일자, 수정일자를 필드로 가지고 있습니다.

보통 테이블에 작성일자, 수정일자는 거의 필수로 들어가는 필드이지만
작성자, 수정자의 경우 들어갈수도 있고 안들어 갈수도 있는 필드죠.

그래서 이 두 부분을 나누고
모든 엔티티에는 BaseTimeEntity를 상속받게 하고
작성자, 수정자가 필요한 엔티티에는 BaseTimeEntity를 상속받은 BaseEntity를 상속받도록 처리합니다.
결과적으로 BaseEntity를 상속받은 엔티티는 작성자, 수정자, 작성일자, 수정일자를 필드로 가지게 되겠죠.

(2) @LastModifiedDate를 이용하면 따로 작성하지 않아도 해당 시간으로 적용이 되는 방식인가요?

네 맞습니다. ㅎㅎ
여러모로 편리한 기능이 많더라구요. ㅋㅋ
부가기능 중에 작성자, 수정자에 대해서 자동으로 넣어주는 기능도 있으니
다음에 이 부분에 대해서 공부해보시면 프로젝트 할 때 편할꺼라 생각합니다! ㅎㅎ

Young-BLUE commented on devrun/src/main/java/kdt/prgrms/devrun/common/aop/GeneralExceptionHandler.java:
```diff
@@ -0,0 +1,54 @@
+@Slf4j
+@RestControllerAdvice
+public class GeneralExceptionHandler {
+
```
Controller에서 @ExceptionHandler 어노테이션이 아닌 따로 class로 빼신 이유가 궁금합니다! 관리하기 편함인가요? 저는 잘 몰라서 기존에 있는 Exception class 들을 그대로 썼는데 추가적인 Exception들을 따로 정의하신 이유가 궁금합니다! :)

Young-BLUE replied:
아하 명확한 예외의 발생 부분을 쉽게 알기 위해서군요. 전에 다른 업무를 할 때 예외를 전체에 씌워서 처리하느라 정확히 어느 부분인지 찾기 어려웠던 코딱지쓰 시절이 생각나네요 감사합니다 🙂

Young-BLUE commented on devrun/src/test/java/kdt/prgrms/devrun/post/controller/PostApiControllerTest.java:
```diff
@@ -0,0 +1,125 @@
+
+    final Long VALID_POST_ID = 1L;
+    final Long INVALID_POST_ID = 1000L;
+
```
이렇게 미리 작성해두면 변수 값으로 새로운 인스턴스 생성에 사용할 수 있군요! 저는 매번 손 아프게 작성했는데 바보였습니다... ㅋㅋㅋ 효율적인 테스트 코딩 배워갑니다! ㅠㅠ

Young-BLUE reviewed (commented):
안녕하세요 지훈님!
제 실력은 이제 갓 첫걸음을 뗀 아기에 불과하지 않는데
리뷰보다는 배워가는 마음으로 리뷰를 해보았습니다 ㅎㅎ
정말 다양한 방법으로 구현할 수 있는 코드는 매력적인거 같습니다.
덕분에 새로운 어노테이션과 구현 방법들을 배워가네요
감사합니다!

🔴🔺🟥
- 당신의 마니또로부터 -

DevRunner21 committed:
[feat] Rest-Docs 라이브러리를 세팅하라.

DevRunner21 committed:
[refactor] 날짜타입을 정해진 포멧의 String으로 반환하도록 수정하라.

DevRunner21 committed:
[refactor] Entity의 제약조건 및 컬럼이름 등을 재설정하라 / PostEntity의 필드를 수정하라

- 각 필드마다 컬럼명을 기입함
- Unique 제약조건 별도 생성 및 명시
- PostEntity에 hobby 필드 제거
- PostEntity에서 createdBy 필드 제거

DevRunner21 committed:
[refactor] Post 조회 시 User와 페치조인하여 한번에 데이터를 가져오도록 수정하라.

DevRunner21 committed:
[refactor] Response용 객체(ApiResult)를 수정하라.

- 전송 시점의 서버시간을 필드로 추가
- status 필드 제거

DevRunner21 committed:
[refactor] Controller의 URI를 전부 복수형으로 수정하라.

DevRunner21 committed:
[refactor] DTO 네이밍을 재정의하라.

DevRunner21 committed:
[refactor] DTO의 Validation 설정을 수정하라.

- Request 객체에 Validation 설정 추가
- Response 객체에 Validation 설정 제거

DevRunner21 committed:
[refactor] DTO의 네이밍을 변경하라. ( *Request -> *RequestDto )

DevRunner21 committed:
[refactor] Page 객체 대신 페이징 데이터용 DTO를 만들고 그것을 사용하라.

DevRunner21 committed:
[refactor] 엔티티의 생성자를 제한하라.

- Builder 로만 생성이 가능하도록 생성자의 접근제한자를 Protected로 설정함

DevRunner21 committed:
[refactor] PostService 테스트 케이스를 수정하라.

- 테스트 케이스 별 메소드 이름 수정
- 테스트 명세를 더 자세하게 수정 ( 테스트 대상, 조건, 반환값이 들어가도록 함 )
- setUp 메소드에 꼭 필요한 데이터만 세팅하도록 함 -> 나머지는 각 테스트 케이스에서 각각 세팅하도록 수정함

DevRunner21 committed:
[refactor] 키 생성 전략을 IDENTITY로 변경하라.

DevRunner21 committed:
[refactor] Controller 테스트 케이스를 수정하라.

DevRunner21 committed:
[feat] ControllerTest에 Rest-Docs 작성 코드를 추가하라.

DevRunner21 committed:
[feat] RestDocs를 작성하라.

DevRunner21 committed:
[refactor] restdocs Gradle 설정을 수정하라.

DevRunner21 committed:
[refactor] 전반적인 테스트 코드를 수정하라.

DevRunner21 committed:
[feat] ServiceLayer에 대한 Request, Response 로깅 기능을 구현하라.

DevRunner21 commented:
안녕하새요! 흑구님!
피드백 주신 내용에 대해 수정한 부분에 대해서 피드백 요청드립니다!

## ✅ 피드백 반영사항  <!-- 지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->

### 반영사항
- [x] Controller URI를 만들때 전부 복수형으로 URL을 만들자. 4efad6559d4c8e30f1f3634ba8f5e878dc421adb
- [x] 모든 필드에 "@Column" 어노테이션을 넣기 ( 자바 소스를 봤을 때 DB 스키마가 보이도록! ) 2bccc54dd5f553ccb1546c2ee7e25e368af1c879
- [x] 유니크 조건을 부여하고 유니크 조건에 대해 네이밍 하기 2bccc54dd5f553ccb1546c2ee7e25e368af1c879
- [x] DTO의 네이밍을 명확하게 하기 ( Request용  Response용 DTO를 각각 만들기 ) 952e508b5963e919cd4ce1cbbdbbfe54a7b0f91c
- [x] 테스트 코드 수정하기. ( setUp() 메서드의 영향을 최소화함 ) 

### 미반영 사항
- [ ] GetMapping 같은 어노테이션에서 아래의 메서드가 어떻게 동작하는지 알아보자.
- [ ] 예외발생 컨트롤을 위한 AssertUtil을 만들어서 예외를 처리하는 방법에 대해 알아보기
- [ ] Spring Security 사용 없이 로그인 로직 만들어 보기.
 
## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
- 코드와 함께 코멘트로 달아놓겠습니다.

DevRunner21 commented on devrun/src/main/java/kdt/prgrms/devrun/common/dto/PageDto.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.List;
+
+@Data
+public class PageDto<T> {
+
+    private long totalCount;
+
+    private int pageNo;
+
+    private int pageSize;
+
+    private List<T> list;
+
+    public PageDto(Page<T> tPage) {
+        this.totalCount = tPage.getTotalElements();
+        this.pageNo = tPage.getNumber();
+        this.pageSize = tPage.getSize();
+        this.list = tPage.getContent();
+    }
+
+}
```
Pageing 데이터를 반환할 때,
JPA에서 제공해주는 Page 객체는 너무 많은 데이터가 들어있어서..
별도 DTO를 만들어서 반환하도록 했습니다.

여기서 저는 이 데이터를 받는 대상은 Web Client 개발자라고 가정하고 API을 만들다보니
페이징 처리가 가능할 정도의 데이터만 내려주도록 필드를 구성했습니다.

이 DTO가 설계상 적절했는지 피드백 부탁드립니다.

DevRunner21 commented on devrun/src/main/java/kdt/prgrms/devrun/domain/User.java:
```diff
@@ -0,0 +1,60 @@
+@AllArgsConstructor(access = AccessLevel.PROTECTED)
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Getter
+@Table(
+    name = "user",
+    uniqueConstraints = {
+    @UniqueConstraint(name = "UniqueLoginId", columnNames = {"login_id"}),
+    @UniqueConstraint(name = "UniqueEmail", columnNames = {"email"}),
+})
```
피드백 주신데로 Unique 제약조건에 대해서 별도 네이밍을 해봤습니다.
일단 구글링하면서 예제로 나왔던 소스들의 네이밍 방법을 따라해봤는데.. 
( 보동 Unique + 적용대상 컬럼명  으로 하더라구요.. )

이 부분에 대해 네이밍이 적절했는지.. 더 보강해야할 부분은 없는지 피드백 부탁드립니다!

DevRunner21 replied:
테이블 명까지 넣어주면 좀 더 출처가 명확해지겠군요!
수정하겠습니다! ㅎㅎ

DevRunner21 commented on devrun/src/test/java/kdt/prgrms/devrun/post/service/SimplePostServiceTest.java:
```diff
@@ -0,0 +1,215 @@
+    @Test
+    @DisplayName("getAllPostPagingList()는 페이지 정보를 파라미터로 받으면, 게시글 페이징 목록을 조회하여 반환합니다.")
+    void test_getAllPostPagingList() {
+
+        // given
+        IntStream.range(0, 30).forEach(i -> postService.createPost(AddPostRequestDto.builder().title("제목 " + i).content("내용").createdBy(user1.getLoginId()).build()));
+
+        final PageRequest pageRequest = PageRequest.of(0, 10);
+
+        // when
+        final PageDto<SimplePostDto> postDtoPage = postService.getPostPagingList(pageRequest);
+
+        // then
+        assertThat(postDtoPage.getTotalCount(), is(30L));
+        assertThat(postDtoPage.getPageNo(), is(0));
+        assertThat(postDtoPage.getPageSize(), is(10));
+        assertThat(postDtoPage.getList(), not(empty()));
+    }
+
+    @Test
+    @DisplayName("getPostById()는 올바른 PostId를 파라미터로 받으면, 게시글의 상세정보를 조회하여 반환합니다.")
+    void test_getPostById_with_validId() {
+
+        // given
+        final Post post1 = Post.builder()
+            .title("제목 1")
+            .content("내용 1")
+            .user(user1)
+            .build();
+        postRepository.save(post1);
+        postRepository.flush();
+
+        final DetailPostDto foundPost = postService.getPostById(post1.getId());
+
+        assertThat(foundPost, notNullValue());
+        assertThat(foundPost.getId(), is(post1.getId()));
+        assertThat(foundPost.getTitle(), is(post1.getTitle()));
+        assertThat(foundPost.getContent(), is(post1.getContent()));
+
+    }
```
이번에 테스트 케이스를 수정하면서
DisplayName에 메서드 설명을 다음과 같이 수정해봤습니다.

테스트코드를 다른 개발자들이 볼 문서라고 생각 했을 때,
테스트 대상 메서드명, 요청 파라미터, 반환값이 들어가야 할 거라고 생각해서 위와 같이 구성해 봤습니다.

테스트 케이스 내용이 적절한지..
단위 테스트에 대한 명세가 적절한지.. 피드백 부탁드립니다.

WooSungHwan replied:
`테스트코드를 다른 개발자들이 볼 문서라고 생각 했을 때,`
이 말씀은 다른 서버 개발자들을 말씀하시는 거겠죠?? 보통 클라이언트는 서버 테스트코드를 보지는 않습니다. 문서로만 소통하고 어느정도 영역을 나누길 원해서..ㅎ

그리고 아마 display에 넣으면 테스트 실행할 때 나오는 메소드 단위에 출력이 될텐데 거기에는 잘릴거에요. 그게 목적이 아니라면 개인적으로는 주석으로 표시해주는걸 선호하는 편입니다.

DevRunner21 commented on devrun/src/main/java/kdt/prgrms/devrun/common/aop/ServiceLoggingAspect.java:
```diff
@@ -0,0 +1,36 @@
+import org.aspectj.lang.annotation.Aspect;
+import org.springframework.stereotype.Component;
+
+@Component
+@Aspect
+@Slf4j
+public class ServiceLoggingAspect {
+
+    @Around(
+        "execution(public * kdt.prgrms.devrun.post..*Service.*(..))"
+//        + " || execution(public * kdt.prgrms.devrun.user..*Service.*(..))"
+        )
+    public Object doLogging(ProceedingJoinPoint pjp) throws Throwable {
+
+        long start = System.currentTimeMillis();
+
+        Object result = null;
+
+        try {
+            log.info("Start [{}] [{}]: Parameter: {}", pjp.getTarget().getClass(), pjp.getSignature().getName(), pjp.getArgs());
+            result = pjp.proceed(pjp.getArgs());
+            return result;
+        } finally {
+            long end = System.currentTimeMillis();
+            log.info("End [{}] [{}]: Return: {} ({}ms)", pjp.getTarget().getClass(), pjp.getSignature().getName(), result, end-start);
+
+        }
```
Spring AOP를 활용해서 ServiceLayer 메서드의 Input, Output에 대한 로깅을 진행했습니다.

일단 로깅 대상 클래스명, 메서드 명, Input/Output 값, API 처리시간 을 넣어봤습니다.
이 부분에 대해서 로깅방식이 적절했는지 피드백 부탁드립니다.

DevRunner21 committed:
[refactor] User 객체의 Builder 메소드를 수정하라.

WooSungHwan commented on devrun/src/test/java/kdt/prgrms/devrun/post/service/SimplePostServiceTest.java:
```diff
@@ -0,0 +1,214 @@
+    void test_getAllPostPagingList() {
+
+        // given
+        IntStream.range(0, 30).forEach(i -> postService.createPost(AddPostRequestDto.builder().title("제목 " + i).content("내용").createdBy(user1.getLoginId()).build()));
+
+        final PageRequest pageRequest = PageRequest.of(0, 10);
```
어쨌든 service 호출 30번을 통해서 프록시 호출이 30번 일어날겁니다. 따라서 다른 영속성 컨텍스트를 갖게 되겠죠? 그렇게되면 테스트 자체의 효율이 너무 떨어집니다.(given으로 인해서..) 그러기보단 given은 테스트를 위한 최소한의 적절한 갯수만을 셋팅을 하고 결과검증에 초점을 맞추는게 어떨까 합니다. limit 10단위로 페이징할건데 30개까지는 필요가 없지 않을까요?

저는 상관없다고 생각해서 이렇게하는데요.
repository 주입받고 saveAll 호출해서 이어나갑니다. 
이에 대해 제가 궁금해서 간단히 테스트해본게 있는데.. https://sas-study.tistory.com/388 한번 구경해보세여 ㅎㅎ
테스트 코드라도 이 코드가 효율적이지 못하다면 리팩토링되어져야 하는건 필요하니까요!ㅎㅎ

WooSungHwan reviewed (approved):
지훈님 코멘트 확인 부탁드립니다 ㅎㅎ

DevRunner21 committed:
[refactor] Asciidoc 설정파일을 수정하라.

- "Unresolved directive in api.adoc - include::"  에러 해결

DevRunner21 committed:
[feat] RestDocs 문서를 생성하라.

DevRunner21 committed:
[refactor] AllArgumentConstructor를 제거하라.

DevRunner21 commented on devrun/src/main/java/kdt/prgrms/devrun/domain/Post.java:
```diff
@@ -0,0 +1,50 @@
+import javax.persistence.*;
+
+@Entity
+@AllArgsConstructor(access = AccessLevel.PROTECTED)
```
아 이 부분 지운다는걸 깜빡하고 그냥 올렸었네요.. ㅜㅜ

원래 "Post의 생성지점을 Builder로 제한하자." 라는 생각을 가지고 있었습니다.
그래서 Builder를 제외하고 생성자에 제약을 걸 생각이었죠.

그래서 처음에 AllArgsConstructor에 아무 생각없이 Protected 접근제한자를 지정했는데..

Protected의 범위 자체가 같은 패키지 내에서는 접근이 허용되다보니
생성지점을 Builder로 한정하려는 제 의도와 맞지 않다고 생각이 들었고

또 그냥 생성자가 없으면 되는거 아닌가? 라는 생각에 도달해서
지금은 생성자 자체를 지웠습니다!

DevRunner21 commented on devrun/src/main/java/kdt/prgrms/devrun/common/enums/ErrorInfo.java:
```diff
@@ -0,0 +1,20 @@
+    POST_NOT_FOUND("POST_NOT_FOUND","게시글을 찾을 수 없습니다."),
+    UNKNOWN("UNKNOWN", "서버 에러로 인해 데이터를 로드 힐 수 없습니다.");
+
+    ErrorInfo(String code, String message) {
```
인텔리제이의 자동완성으로 생성자를 만들다보니 체크를 제대로 못한 것 같습니다.
Enum의 경우 인텔리제이에서 자동완성으로 생성자를 만들면 접근제어자가 안들어가는군요.. ㅜㅜ

이번에 구글링해서 안 사실인데..
Enum 타입의 생성자는 접근제한자가 무조건 private라고 합니다!

Enum 자체가 고정된 상수들의 집합니다 보니
런타임이 아닌 컴파일 타임에 모든 값이 세팅되어야 하고
런타임에 그 값이 다른 곳에서 변경되면 안되다보니
생성자의 접근제어자를 private로 지정해놓았다고 합니다. 

흑구님 덕분에  또 한가지 사실을 알아가는군요! ㅎㅎ

DevRunner21 committed:
[refactor] ErrorEnum에 소스 오타를 수정하라.

- 생성자에 접근제한자 붙이기
- 메시지 오타 수정

DevRunner21 commented on devrun/src/main/java/kdt/prgrms/devrun/post/service/PostService.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service
```
지우는걸 깜빡한 소스입니다. ㅜㅜ
앞으로는 푸시하기전에 소스를 좀 더 꼼꼼히 점검해봐야겠습니다.ㅜㅜ
알려주셔서 감사합니다! ㅜㅜ

DevRunner21 committed:
[refactor] 불필요한 소스를 제거하라.

DevRunner21 committed:
[refactor] convert 메서드에 생성로직을 builder를 사용하도록 통일하라.

DevRunner21 committed:
[refactor] DTO에도 Entity 처럼 길이 제약조건을 정의하라.

DevRunner21 committed:
[refactor] PageDto에 hasNext 필드를 추가하라.

DevRunner21 committed:
[refactor] Unique제약조건 이름을 수정하라.

DevRunner21 committed:
[refactor] 외래키 제약조건 이름을 지정하라.

DevRunner21 committed:
[refactor] 테스트 코드에서 가데이터 주입부분을 수정하라.

- save()에서 saveAll()을 사용하도록 변겅

DevRunner21 committed:
[refactor] Controller에 MediaType을 설정하라.

DevRunner21 committed:
[feat] 사용자 회원가입 기능을 구현하라. ( 암호화 미적용 )

- 사용자 등록 기능 구현
- 사용자 ID 중복확인 기능 구현
- 사용자 Email 중복확인 기능 구현

DevRunner21 committed:
[refactor] 사용자 등록 시 등록된 사용자 ID를 반환하라.

DevRunner21 committed:
[feat] 사용자 상세조회 기능을 구현하라.

DevRunner21 committed:
[test] 회원가입 테스트 코드를 작성하라.

- 회원 단건조회 테스트
- 회원가입 테스트
- 회원가입시 중복된 ID 입력 테스트
- 회원가입시 중복된 Email 입력 테스트
